q2cli/cache.py

```diff
@@ -0,0 +1,281 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+
+
+class DeploymentCache:
+    """Cached CLI state for a QIIME deployment.
+
+    In this context, a QIIME deployment is the set of installed Python
+    packages, including their exact versions, that register one or more QIIME 2
+    plugins. The exact version of q2cli is also included in the deployment.
+
+    The deployment cache stores the current deployment's package names and
+    versions in a requirements.txt file under the cache directory. This file is
+    used to determine if the cache is outdated. If the cache is determined to
+    be outdated, it will be refreshed based on the current deployment state.
+    Thus, adding, removing, upgrading, or downgrading a plugin package or q2cli
+    itself will trigger a cache refresh.
+
+    Two mechanisms are provided to force a cache refresh. Setting the
+    environment variable Q2CLIDEV to any value will cause the cache to be
+    refreshed upon instantiation. Calling `.refresh()` will also refresh the
+    cache. Forced refreshing of the cache is useful for plugin and/or q2cli
+    developers who want their changes to take effect in the CLI without
+    changing their package versions.
+
+    Cached CLI state is stored in a state.json file under the cache directory.
+    It is not a public file format and it is not versioned. q2cli is included
+    as part of the QIIME deployment so that the cached state can always be read
+    (or recreated as necessary) by the currently installed version of q2cli.
+
+    This class is intended to be a singleton because it is responsible for
+    managing the on-disk cache. Having more than one instance managing the
+    cache has the possibility of two instances clobbering the cache (e.g. in a
+    multithreaded/multiprocessing situation). Also, having a single instance
+    improves performance by only reading and/or refreshing the cache a
+    single time during its lifetime. Having two instances could, for example,
+    trigger two cache refreshes if Q2CLIDEV is set. To support these use-cases,
+    a module-level `CACHE` variable stores a single instance of this class.
+
+    """
+
+    # Public API
+
+    def __init__(self):
+        import os
```

Comment:
I guess there is some overhead if the function is called repeatedly, even though the imports are cached. It's probably a wash for this one (because this gets instantiated exactly one time), but I'll move it out (I don't have much of an opinion on this one).